### PR TITLE
Warnings as errors compiler build flag

### PR DESCRIPTION
**Pull request by itsgreggreg**
Adds a `--warnings-as-errors` flag to the `build` command. Which optionally turns compile time warnings into compile time errors.

Closes https://github.com/gleam-lang/gleam/issues/967

Side question: Is there a reason why the `build` command is hidden from the `help` message?

### CODE REVIEW

**Review by GentileFulvio — COMMENTED** [review]
I noticed that there aren't any tests for the added functionality ? I could be wrong though.

Sorry for this bold comment but I was looking at the issue and noticed that a PR was created and wondered how one would add this functionality to the code.

**Comment by itsgreggreg** [comment]
@GentileFulvio good lookin out. I'm not too sure how to test this. The existing tests are for functionality deeper than `main.rs`. Maybe Gleam needs public interface tests?

**Comment by lpil** [comment]
We could take an approach like the core language tests and add a CI step that runs on a project in `test` with some warnings in it and assert that it returns non-zero.

We could probably do some some form of CLI level integration testing, but it's not been hugely cost effective to do this yet.

**Comment by lpil** [comment]
> Side question: Is there a reason why the build command is hidden from the help message?

People keep getting confused by it, thinking it is a replacement for rebar3. It's going to be deprecated in a few versions time and be replaced with the new `gleam compile-package` command and the new Gleam build tool.

File: src/error.rs [hunk 1]
```diff
@@ -160,6 +160,10 @@ pub enum Error {
     MetadataDecodeError {
         error: Option<String>,
     },
+
+    GenericFileDiagnostics {
```
Comment (lpil) [inline src/error.rs:164]: I think if we pre-render the warnings and put them here we won't be able to have multiple formats for errors later (i.e. LSP, JSON, GitHub Actions diagnostic thingies). The preciseness of the warnings previously worked well for this.

Perhaps an approach like this?

```rust
    let any_warnings = warning::print_all(analysed.as_slice());
    if warnings_as_errors && any_warnings {
        return Err(Error::ForbiddenWarnings);
    }
```
```
error: Warnings detected

The warnings above were found when compiling this project and the `--warnings-as-errors`
flag has been set so we can't continue compilation. Fix the warnings and try again.
```

**Comment by GentileFulvio** [comment]
> We could take an approach like the core language tests and add a CI step that runs on a project in `test` with some warnings in it and assert that it returns non-zero.
> 
> We could probably do some some form of CLI level integration testing, but it's not been hugely cost effective to do this yet.

And testing the `as_errors` function in `src/warning.rs` is not possible as well I guess ? Because of the `&[crate::project::Analysed]` type right ?

**Comment by lpil** [comment]
We could construct that type and test it if we wanted it. It's just about deciding what we want to test. Currently we don't have any tests around the format that errors get printed with

**Review by lpil — APPROVED** [review]
Very nice, thank you

**Comment by lpil** [comment]
I've rebased these onto main (had to fix a merge conflict). Thanks!
